semgrep-core/tests/js/cp_undeclared.js

```diff
@@ -0,0 +1,3 @@
+foo = "undeclared variable"
+//ERROR:
+console.log(foo)
```

Comment:
Will this also work in a function like so: https://semgrep.dev/s/squaresurf:eslint_detect-object-injection?

Reply:
+1, I tried 

```
value = obj["..."]
```

and then 

```
function x() {
    foo = "undeclared variable"
    //ERROR:
    value = obj[foo]
}
```

and didn't get results, but did with just 

```
foo = "undeclared variable"
//ERROR:
value = obj[foo]
```

Oddly, this doesn't show up when you use

```
function x() {
   foo = "undeclared variable"
   //ERROR:
   console.log(foo)
}
```

Reply:
> Oddly, this doesn't show up when you use ...

Hmm, I'm not sure I follow. I wouldn't expect that to match `value = obj["..."]` because there's no `obj`, or `value`?

Reply:
There is something weird indeed, I'll take a look. Thanks both!

Reply:
oops, @mschwager to clarify, with the test case Iago has

```
   foo = "undeclared variable"
   //ERROR:
   console.log(foo)
```

matches as expected, and so does

```
function x() {
   foo = "undeclared variable"
   //ERROR:
   console.log(foo)
}
```

Reply:
There were several problems at play, first `foo` was incorrectly resolved as a local variable instead of a global, and then the dataflow constant propagation didn't propagate constants inside nested lvalues. Thanks for catching this!